References
======================

* `R. Bonneau et al., “The Inferelator: an algorithm for learning parsimonious regulatory networks from systems-biology data sets de novo,” Genome Biology, vol. 7, p. R36, May 2006. <https://doi.org/10.1186/gb-2006-7-5-r36>`_

* `A. Madar, A. Greenfield, H. Ostrer, E. Vanden-Eijnden, and R. Bonneau, “The inferelator 2.0: A scalable framework for reconstruction of dynamic regulatory network models,” in 2009 Annual International Conference of the IEEE Engineering in Medicine and Biology Society, Sep. 2009. <https://ieeexplore.ieee.org/document/5334018>`_

* `A. Madar, A. Greenfield, E. Vanden-Eijnden, and R. Bonneau, “DREAM3: Network Inference Using Dynamic Context Likelihood of Relatedness and the Inferelator,” PLOS ONE, vol. 5, no. 3, p. e9803, Mar. 2010. <https://doi.org/10.1371/journal.pone.0009803>`_

* `A. Greenfield, A. Madar, H. Ostrer, and R. Bonneau, “DREAM4: Combining Genetic and Dynamic Information to Identify Biological Networks and Dynamical Models,” PLOS ONE, vol. 5, no. 10, p. e13397, Oct. 2010. <https://doi.org/10.1371/journal.pone.0013397>`_

* `M. Ciofani et al., “A Validated Regulatory Network for Th17 Cell Specification,” Cell, vol. 151, no. 2, pp. 289–303, Oct. 2012. <https://doi.org/10.1016/j.cell.2012.09.016>`_

* `A. Greenfield, C. Hafemeister, and R. Bonneau, “Robust data-driven incorporation of prior knowledge into the inference of dynamic regulatory networks,” Bioinformatics, vol. 29, no. 8, pp. 1060–1067, Apr. 2013. <https://doi.org/10.1093/bioinformatics/btt099>`_

* `M. L. Arrieta‐Ortiz et al., “An experimentally supported model of the Bacillus subtilis global transcriptional regulatory network,” Molecular Systems Biology, vol. 11, no. 11, p. 839, Nov. 2015. <https://doi.org/10.15252/msb.20156236>`_

* `O. Wilkins et al., “EGRINs (Environmental Gene Regulatory Influence Networks) in Rice That Function in the Response to Water Deficit, High Temperature, and Agricultural Environments,” The Plant Cell, vol. 28, no. 10, pp. 2365–2384, Oct. 2016. <https://doi.org/10.1105/tpc.16.00158>`_

* `K. Tchourine, C. Vogel, and R. Bonneau, “Condition-Specific Modeling of Biophysical Parameters Advances Inference of Regulatory Networks,” Cell Reports, vol. 23, no. 2, pp. 376–388, Apr. 2018. <https://doi.org/10.1016/j.celrep.2018.03.048>`_

* `D. M. Castro, N. R. de Veaux, E. R. Miraldi, and R. Bonneau, “Multi-study inference of regulatory networks for more accurate models of gene regulation,” PLOS Computational Biology, vol. 15, no. 1, p. e1006591, Jan. 2019. <https://doi.org/10.1371/journal.pcbi.1006591>`_

* `E. R. Miraldi et al., “Leveraging chromatin accessibility for transcriptional regulatory network inference in T Helper 17 Cells,” Genome Res., vol. 29, no. 3, pp. 449–463, Mar. 2019. <http://www.genome.org/cgi/doi/10.1101/gr.238253.118>`_

* `C. A. Jackson, D. M. Castro, G.-A. Saldi, R. Bonneau, and D. Gresham, “Gene regulatory network reconstruction using single-cell RNA sequencing of barcoded genotypes in diverse environments,” eLife, vol. 9, p. e51254, Jan. 2020. <https://doi.org/10.7554/eLife.51254>`_

* `C. S. Gibbs, C. A. Jackson, G.-A. Saldi, A. Tjärnberg, A. Shah, et al. 2022. “High Performance Single-Cell Gene Regulatory Network Inference at Scale: The Inferelator 3.0.” Bioinformatics , February. <https://doi.org/10.1093/bioinformatics/btac117>``